### PR TITLE
Make ctc topo normal.

### PR DESCRIPTION
**Pull request by yaguanghu**
#77 
@danpovey @csukuangfj 
Please have a look at it.

### CODE REVIEW

File: snowfall/training/ctc_graph.py [hunk 1]
```diff
@@ -15,8 +15,8 @@ def build_ctc_topo(tokens: List[int]) -> k2.Fsa:
     symbols to a single output symbol.
 
     Caution:
-      The resulting topo is an FST. Epsilons are on the left
-      side (i.e., ilabels) and tokens are on the right side (i.e., olabels)
+      The resulting topo is an FST. Epsilons are on the right
```
Comment (qindazhu) [inline snowfall/training/ctc_graph.py:18]: It would be better if we can sync the documentation with PR https://github.com/k2-fsa/k2/pull/602/files where @danpovey and I left some comments and @csukuangfj has fixed them.

Comment (yaguanghu) [reply snowfall/training/ctc_graph.py:18]: Sure, I will modify it when @csukuangfj's PR is ready.

File: egs/aishell/asr/simple_v1/decode.py [hunk 2]
```diff
@@ -137,8 +137,8 @@ def main():
     lang_dir = Path('data/lang_nosp')
     symbol_table = k2.SymbolTable.from_file(lang_dir / 'words.txt')
     phone_symbol_table = k2.SymbolTable.from_file(lang_dir / 'phones.txt')
-    ctc_topo = build_ctc_topo(list(phone_symbol_table._id2sym.keys()))
-    ctc_topo = k2.arc_sort(ctc_topo)
+    ctc_topo = k2.arc_sort(build_ctc_topo(list(
```
Comment (csukuangfj) [inline egs/aishell/asr/simple_v1/decode.py:140]: Please update it like the one in librispeech, which excludes disambiguation symbols from 
`phone_symbol_table`. 

Sorry that aishell has not been kept in sync with librispeech. We have been doing
experiments with the librispeech recipe.

Comment (yaguanghu) [reply egs/aishell/asr/simple_v1/decode.py:140]: will do that.

Comment (csukuangfj) [reply egs/aishell/asr/simple_v1/decode.py:140]: Like the one in
https://github.com/k2-fsa/snowfall/blob/1567a6553af8519e52fa86a2c6fd5aa6bf0c43f9/egs/librispeech/asr/simple_v1/mmi_bigram_decode.py#L210

File: snowfall/decoding/graph.py [hunk 3]
```diff
@@ -19,7 +19,7 @@ def compile_LG(L: Fsa, G: Fsa, ctc_topo_inv:Fsa, labels_disambig_id_start: int,
         G:
             An ``Fsa`` that represents the language model (G), i.e. it's an acceptor
             with words as ``symbols``.
-        ctc_topo_inv:  Epsilons are in `aux_labels` and `labels` contain phone IDs.
+        ctc_topo:  Epsilons are in `labels` and `aux_labels` contain phone IDs.
```
Comment (csukuangfj) [inline snowfall/decoding/graph.py:22]: I think `epsilons` are in `aux_labels`.

File: snowfall/training/ctc_graph.py [hunk 4]
```diff
@@ -69,8 +69,7 @@ def __init__(self,
         self.phones = phones
         self.words = words
         self.oov = oov
-        ctc_topo_inv = build_ctc_topo(list(phones._id2sym.keys())).invert_()
-        self.ctc_topo_inv = k2.arc_sort(ctc_topo_inv)
+        self.ctc_topo = k2.arc_sort(build_ctc_topo(list(phones._id2sym.keys())))
```
Comment (csukuangfj) [inline snowfall/training/ctc_graph.py:72]: Please also fix this one. Disambiguation symbols should be excluded.

File: snowfall/decoding/graph.py (hunk 3 above)
Comment (qindazhu) [inline snowfall/decoding/graph.py:22]: I'm confused by this line, I was thinking the left side, i.e. labels,  would contain `blank` instead of `epsilon`

File: snowfall/training/mmi_graph.py [hunk 5]
```diff
@@ -106,7 +106,7 @@ def __init__(self,
         ctc_topo = build_ctc_topo(phone_symbols_with_blank)
         assert ctc_topo.requires_grad is False
 
-        self.ctc_topo = ctc_topo
+        self.ctc_topo_inv = ctc_topo.invert_()
```
Comment (csukuangfj) [inline snowfall/training/mmi_graph.py:109]: Maybe you need to do `k2.arc_sort(ctc_topo.invert_())`.

File: egs/aishell/asr/simple_v1/decode.py [hunk 6]
```diff
@@ -137,8 +137,9 @@ def main():
     lang_dir = Path('data/lang_nosp')
     symbol_table = k2.SymbolTable.from_file(lang_dir / 'words.txt')
     phone_symbol_table = k2.SymbolTable.from_file(lang_dir / 'phones.txt')
-    ctc_topo = build_ctc_topo(list(phone_symbol_table._id2sym.keys()))
-    ctc_topo = k2.arc_sort(ctc_topo)
+    phone_ids = get_phone_symbols(phone_symbol_table)
```
Comment (csukuangfj) [inline egs/aishell/asr/simple_v1/decode.py:140]: Could you also update *train.py* to exclude disambiguation symbols while
building the model?

https://github.com/k2-fsa/snowfall/blob/add948544da04ac4bb9c14cef0057c6969c1b9fb/egs/librispeech/asr/simple_v1/train.py#L298

https://github.com/k2-fsa/snowfall/blob/add948544da04ac4bb9c14cef0057c6969c1b9fb/egs/aishell/asr/simple_v1/train.py#L300

The correct one should look like the following:
https://github.com/k2-fsa/snowfall/blob/add948544da04ac4bb9c14cef0057c6969c1b9fb/egs/librispeech/asr/simple_v1/mmi_bigram_train.py#L334-L336

Comment (yaguanghu) [reply egs/aishell/asr/simple_v1/decode.py:140]: sure.

**Comment by csukuangfj** [comment]
Could you verify that the changes do not break the training and decoding?

**Comment by yaguanghu** [comment]
I don't know how to verify it. Do you have any reasonable results for reference?

**Comment by csukuangfj** [comment]
> I don't know how to verify it.

It can be verified by executing `run.sh`. It should give you a WER at the end. For librispeech, the value is between 10.0% and 11.0%.

If you have run it before,
- For the training part, it should converge like your previous runs
- For the decoding part, it should give the same WER as your previous checkpoint

**Comment by qindazhu** [comment]
You can just run `run.sh` (by default it runs`mmi_bigram_*`, so you may need to run `train/decode.py` seperately), the WER of decoding should be in 10%~12%. Noted the WER range is for LibriSpeech, I never run experiments for aishell with latest k2 changes, sorry.

BTW, it would be great if we can rename `train/decode.py` to `ctc_train.py` and `ctc_test.py`

**Comment by yaguanghu** [comment]
OK, I will verify the results and refine the training scripts.

**Comment by yaguanghu** [comment]
I have run the training and decoding pipeline, the final WER is 13.59%. It seems a little worse than your result.
I noticed that the default training script uses only train-clean-100 for training, what's your training data?

**Comment by csukuangfj** [comment]
We have been using run.sh and the same settings in prepare.py.  

Thursday, 21 January 2021, 20:00 +0800 from notifications@github.com  <notifications@github.com>:

**Comment by danpovey** [comment]
It uses musan, right?  And we are talking about the librispeech example?

**Comment by danpovey** [comment]
Possibly there is just a big scatter of results.  If so we should
definitely investigate the reasons.
What I'd like to see is a comparison of the norms of different parameter
matrices of the trained models.
If a particular layer is getting unstable, it may reveal it.
IDK what diagnostics are available from the logs we automatically write?
E.g. convergence plots?

**Comment by yaguanghu** [comment]
I ran the whole pipeline again, and the final WER is 13.23%. I think the first thing we need to do is eliminating randomness, so that the final results is trustable.
BTW I have some problems in running the original pipeline, but I will figure it out and make a comparsion with this change.

**Comment by danpovey** [comment]
Did you run
python3 setup.py install
before running?

**Comment by yaguanghu** [comment]
I ran `pip install -e .`

File: snowfall/training/ctc_graph.py [hunk 7]
```diff
@@ -69,8 +67,9 @@ def __init__(self,
         self.phones = phones
         self.words = words
         self.oov = oov
-        ctc_topo_inv = build_ctc_topo(list(phones._id2sym.keys())).invert_()
-        self.ctc_topo_inv = k2.arc_sort(ctc_topo_inv)
+        phone_ids = get_phone_symbols(phones)
```
Comment (danpovey) [inline snowfall/training/ctc_graph.py:70]: A difference here is that ctc_topo no longer includes the disambig symbols.
Of course it's not right to treat the disambig symbols as if they were phones.
But doesn't L contain disambig symbols on its input side?  Perhaps what is happening is that we are eliminating paths with disambig symbols.   We might need to add self-loop with each of the disambig symbols, on all of the states of the ctc topo.

File: snowfall/training/ctc_graph.py [hunk 8]
```diff
@@ -85,9 +84,10 @@ def compile_one_and_cache(self, text: str) -> k2.Fsa:
         tokens = (token if token in self.words else self.oov
                   for token in text.split(' '))
         word_ids = [self.words[token] for token in tokens]
-        fsa = k2.linear_fsa(word_ids)
-        decoding_graph = k2.connect(k2.intersect(fsa, self.L_inv)).invert_()
+        label_graph = k2.linear_fsa(word_ids)
+        decoding_graph = k2.connect(k2.intersect(label_graph,
+                                                 self.L_inv)).invert_()
```
Comment (danpovey) [inline snowfall/training/ctc_graph.py:89]: Actually, instead of adding the disambig symbols into the ctc_topo, we can handle them here.
  (i) we can use compose instead of intersection and double inversion, here.
  (ii) after intersecting, we can replace all the disambig symbols with epsilon and then remove epsilons, and then connect.

Comment (danpovey) [reply snowfall/training/ctc_graph.py:89]: Oh, wait... now I see that train.py is loading L.fst not L_disambig.fst.  So there may be no disambig symbols to deal with.

**Comment by danpovey** [comment]
I suspect there is a bug but don't know what it is.  Perhaps we could start by printing the compiled CTC training graphs for the first minibatch, for this and the unmodified version, and see if there is a difference.
And same with the decoding graph.  (Careful, it is cached by default).

File: snowfall/training/ctc_graph.py (hunk 7 above)
Comment (yaguanghu) [reply snowfall/training/ctc_graph.py:70]: Now the L graph contain disambig symbols on its input side. That might be a problem in the current ctc topology. I think it makes sense to add  self-loop with each disambig symbols.

Comment (danpovey) [reply snowfall/training/ctc_graph.py:70]: I think I was mistaken about this L having disambig symbols, we are loading L.fst in train.py, which lacks them.

Comment (yaguanghu) [reply snowfall/training/ctc_graph.py:70]: I will check it. If there are no disambig symbols in L.fst, there may be problems while decoding.

**Comment by danpovey** [comment]
We use L_disambig.fst to create the decoding graph.  But the disambig
symbols are removed prior to composing with ctc_topo.
Bear in mind that it should be possible to train a model with one version
of the code and decode with another; that will help narrow
the problem down to training or decoding.

**Comment by csukuangfj** [comment]
I started to re-run the LibriSpeech recipe from scratch using the master branch this morning. Here is the WER

```
2021-01-22 11:49:11,000 INFO [mmi_bigram_decode.py:296] %WER 10.35% [5443 / 52576, 772 ins, 479 del, 4192 sub ]
```

**Comment by yaguanghu** [comment]
> I started to re-run the LibriSpeech recipe from scratch using the master branch this morning. Here is the WER
> 
> ```
> 2021-01-22 11:49:11,000 INFO [mmi_bigram_decode.py:296] %WER 10.35% [5443 / 52576, 772 ins, 479 del, 4192 sub ]
> ```

Aren't we talking about CTC training? My result is based on CTC pipeline.

**Comment by danpovey** [comment]
Ah, that may explain it, does anyone remember what the CTC WER should be?  I would have thought it would be a little better than that, especially now that we have augmentation.

We should test the MMI branch too, because that also uses the ctc topo.  However, I think it would make more sense to modify mmi_graph.py so that it stores ctc_topo instead of ctc_topo_inv, and uses compose instead of intersect then invert.

**Comment by csukuangfj** [comment]
From https://github.com/k2-fsa/snowfall/pull/57#issuecomment-750297209

> 2020-12-23 20:09:55,264 INFO [decode.py:214] %WER 12.97% [6817 / 52576, 897 ins, 654 del, 5266 sub ]

It was trained without MUSAN.

**Comment by danpovey** [comment]
Ah OK.
Can we start putting the WERs at the top of the training scripts?
(mentioning which decode script we used, and when, who ran it,
what changed if there are multiple WERs listed, etc.)?
Yaguang, were you using MUSAN?
I assume yes, if you ran the whole thing from scratch.
Dan

**Comment by yaguanghu** [comment]
It seems like the default ctc training script is now using musan for data augmentation, althoug it prepared musan data.
I will run it again with musan augmentation.

**Comment by csukuangfj** [comment]
WER of CTC training from scratch using the master branch

```
2021-01-22 16:07:01,722 INFO [decode.py:214] %WER 12.76% [6709 / 52576, 901 ins, 747 del, 5061 sub ]
```

**Comment by csukuangfj** [comment]
> It seems like the default ctc training script is now using musan for data augmentation

I have been assuming that your changes are based on the latest master and your code is using MUSAN to verify
the changes.

**Comment by yaguanghu** [comment]
WER of CTC training with musan using this change.
`INFO [ctc_decode.py:216] %WER 12.47% [6555 / 52576, 802 ins, 708 del, 5045 sub ]`

**Comment by danpovey** [comment]
great!we.can merge?  fangjun pls merge if ok w.you.

**Comment by yaguanghu** [comment]
Add the result of master branch ran by me, note this result is trained without musan, I'm also training CTC with musan using master branch.
`INFO [decode.py:214] %WER 13.21% [6946 / 52576, 909 ins, 705 del, 5332 sub ]`

**Comment by csukuangfj** [comment]
+2

Friday, 22 January 2021, 17:26 +0800 from notifications@github.com  <notifications@github.com>:

**Comment by xiaohui-zhang** [comment]
> Add the result of master branch ran by me, note this result is trained without musan, I'm also training CTC with musan using master branch.
> `INFO [decode.py:214] %WER 13.21% [6946 / 52576, 909 ins, 705 del, 5332 sub ]`

Hey Yaguang, can you add this CTC result (or the one above, whichever should be re-produced by librispeech/run.sh w/ CTC's train.py) to the RESULTS file, before the most recent MMI+MBR numbers? Maybe add MMI numbers also if you it's easy for you to get. Keeping track of these will make debug easier in future.
